File: backend/src/services/perfilService.js
const { IllegalArgumentError } = require('../middleware/errorHandler');
const { roleFirestoreParaPerfil } = require('../utils/roleUtils');

class PerfilService {
  constructor(firestoreService) {
    this.firestoreService = firestoreService;
  }

  async resolverPerfil(uid, email) {
    const doc = await this.buscarDocumentoUsuario(uid, email);
    const role = doc.get('role');
    return this.normalizarRole(role);
  }

  async resolverPerfilCompleto(uid, email) {
    const doc = await this.buscarDocumentoUsuario(uid, email);
    const role = doc.get('role');

    return {
      uid: doc.id,
      email: doc.get('email') || email,
      nome: doc.get('nome') || '',
      perfil: this.normalizarRole(role)
    };
  }

  async buscarDocumentoUsuario(uid, email) {
    const byUid = await this.firestoreService.buscarUsuarioPorUid(uid);
    const doc = byUid || (await this.firestoreService.buscarUsuarioPorEmail(email));
    if (!doc) throw new IllegalArgumentError('Usuario nao encontrado no banco de dados.');
    return doc;
  }

  normalizarRole(role) {
    if (!role || !String(role).trim()) throw new IllegalArgumentError('Usuario sem role configurada no banco.');
    const perfil = roleFirestoreParaPerfil(role);
    if (!['admin', 'coordenador', 'aluno'].includes(perfil)) {
      throw new IllegalArgumentError('Role invalida no banco: ' + role);
    }
    return perfil;
  }
}

module.exports = PerfilService;

File: backend/app/services/perfil_service.rb
# frozen_string_literal: true

class PerfilService
  def initialize(firestore_service: FirestoreService.new)
    @firestore = firestore_service
  end

  def resolver_perfil(uid, email)
    doc = @firestore.buscar_usuario(uid, email)
    raise ArgumentError, "Usuario nao encontrado no banco de dados." unless doc

    normalizar_role(valor_perfil(doc))
  end

  private

  def valor_perfil(doc)
    data = doc[:data]
    FirestoreKeyNormalizer.fetch_field(data, "role", "perfil", "tipo")
  end

  def normalizar_role(role)
    raise ArgumentError, "Usuario sem role configurada no banco." if role.blank?

    normalizado = FirestoreKeyNormalizer.normalize_role_for_db(role)
    return "admin" if normalizado == "superadmin"

    unless %w[coordenador aluno admin].include?(normalizado)
      raise ArgumentError, "Role invalida no banco: #{role}"
    end

    normalizado
  end
end
